smart-contracts/contracts/Airdrops.sol

```diff
@@ -0,0 +1,187 @@
+   * @param _token The address of the ERC20 token contract (UP).
+   */
+  constructor(address _token) {
+    token = IERC20(_token);
+    owner = msg.sender;
+  }
+
+  /**
+   * @notice Modifier to restrict functions to the contract owner.
+   */
+  modifier onlyOwner() {
+    if (msg.sender != owner) revert OnlyOwner();
+    _;
+  }
+
+  /**
+   * @notice Sets the Merkle root and Terms of Service (TOS) for a campaign.
+   * @param campaignName The name/identifier of the campaign.
+   * @param tos The Terms of Service text for the campaign.
+   * @param root The Merkle tree root corresponding to eligible entries.
+   */
+  function setMerkleRootForCampaign(
+    string calldata campaignName,
+    string calldata tos,
+    bytes32 root
+  ) external override onlyOwner {
+    campaigns[campaignName] = Campaign(tos, root);
+    emit CampaignSet(campaignName, root);
+  }
+
+  /**
+   * @notice Signs the Terms of Service for a given campaign.
+   *
+   * The signature must be produced off-chain, signing the Ethereum Signed Message hash of the campaign's TOS.
+   *
+   * @param campaignName The campaign identifier.
+   * @param recipient The address that is signing the TOS.
+   * @param signature The signature bytes (generated off-chain) of the hash of the campaign's TOS.
+   */
+  function signTos(
+    string calldata campaignName,
+    address recipient,
+    bytes calldata signature
+  ) external override {
+    if (msg.sender != recipient) revert NOT_AUTHORIZED();
+
+    Campaign storage campaign = campaigns[campaignName];
+    require(bytes(campaign.tos).length != 0, "Campaign does not exist");
+
+    // Create the Ethereum Signed Message hash from the campaign's TOS.
+    bytes32 messageHash = MessageHashUtils.toEthSignedMessageHash(keccak256(bytes(campaign.tos)));
+    // Recover the signer from the provided signature.
+    address recovered = ECDSA.recover(messageHash, signature);
+    if (recovered != recipient) {
+      revert NOT_AUTHORIZED();
+    }
+    signedTos[campaignName][recipient] = true;
+    emit TosSigned(campaignName, recipient);
+  }
+
+  /**
+   * @notice Claims tokens for the airdrop if the recipient is eligible via the Merkle proof and has signed the TOS.
+   *
+   * It computes the leaf from the recipient address and token amount using the same logic as the backend:
+   * 
+   * \[
+   * \text{leaf} = \mathtt{keccak256( bytes.concat( keccak256( abi.encode( recipient, amount ) ) ) )}
+   * \]
+   *
+   * @param campaignName The campaign identifier.
+   * @param recipient The address claiming tokens. (Must be equal to msg.sender.)
+   * @param amount The token amount to claim.
+   * @param proof The concatenated proof bytes (multiple of 32 bytes) used for Merkle proof verification.
+   */
+  function claim(
+    string calldata campaignName,
+    address recipient,
+    uint256 amount,
+    bytes calldata proof
+  ) external override {
+    if (msg.sender != recipient) revert NOT_AUTHORIZED();
```

Comment:
Not needed, someone could claim for someone else... As long as they have signed the tos